File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.py
def solution(id_list, reports, k):
    dic = {}
    # [내가 신고한 사람], [나를 신고한 사람]

    for id in id_list:
        dic[id] = [[], []]

    for report in reports:
        report_from, report_to = report.split()
        if report_from in dic:
            if report_to not in dic[report_from][0]:
                dic[report_from][0].append(report_to)

        if report_to in dic:
            if report_from not in dic[report_to][1]:
                dic[report_to][1].append(report_from)
    answer = []
    for id in id_list:
        cnt = 0
        report_list = dic[id][0]
        for r in report_list:
            if len(dic[r][1]) >= k:
                cnt += 1
        answer.append(cnt)

    return answer

File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.java
import java.util.*;

class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        // 신고당한 횟수 초기화
        Map<String, Integer> reported = new HashMap<>();
        for (String id : id_list) {
            reported.put(id, 0);
        }

        // 신고 기록 초기화
        Map<String, List<String>> reportLog = new HashMap<>();
        for (String id : id_list) {
            reportLog.put(id, new ArrayList<>());
        }

        // 신고 처리
        for (String r : report) {
            String[] split = r.split(" ");

            String from = split[0];
            String to = split[1];

            List<String> reportList = reportLog.get(from);
            if (!reportList.contains(to)) {
                reportList.add(to);
                reported.put(to, reported.get(to) + 1);
            }
        }

        int[] result = new int[id_list.length];
        for (int i=0; i<id_list.length; i++) {
            int count = 0;
            for (String r : reportLog.get(id_list[i])) {
                if (reported.get(r) >= k) {
                    count += 1;
                }
            }
            result[i] = count;
        }

        return result;
    }
}